### ISSUE TITLE
`LoRAConverter` silently leaves non-Linear parameters trainable

### ISSUE THREAD

**Issue by yh8899** [pytorch/torchtitan#3438]
### Summary

`torchtitan.components.lora.LoRAConverter` only walks `Linear.Config`
nodes in the model config tree — target Linears become `LoRALinear`
(base frozen), non-target Linears get wrapped with `FrozenConfig`.
Every **non-Linear** trainable parameter (RMSNorm `weight`, embedding
`weight`) keeps the default `requires_grad=True`, so the default
optimizer (which filters by `if p.requires_grad`) starts training them
alongside the LoRA adapters.

The API is named `LoRAConverter` and is the only obvious LoRA on-ramp,
so users (and the project's own integration test
`llama3_debugmodel_float8_emulate_lora`) end up training norms and
embeddings without realising.

### Why it matters

For each step, the leaked non-Linear params:

1. Get **optimizer steps + Adam state** — wasted memory and compute.
2. Get **FSDP reduce-scatter** for their (effectively zero) gradient
   buckets — wasted communication.
3. **Diverge from the pretrained values** silently — there's no
   user-visible signal that "LoRA training" is also drifting
   normalisation statistics and token embeddings.

For LoRA-only checkpoint flows (downstream code that stores just
`.lora_a.*` / `.lora_b.*`), the accidentally-trained
norms/embeddings **aren't saved**, so every resume reloads the HF
base and wipes the accumulated updates — the wasted training was
strictly negative.

### Reproduction

```python
"""Uses the same flavor + target_modules as torchtitan's own LoRA
integration test ``llama3_debugmodel_float8_emulate_lora`` in
``torchtitan/models/llama3/config_registry.py``.
"""
import torch
from torchtitan.components.lora import LoRAConverter
from torchtitan.models.llama3 import llama3_configs

cfg = llama3_configs["debugmodel"](attn_backend="flex")
LoRAConverter.Config(
    rank=8, alpha=16.0, target_modules=["wq", "wkv", "wo"]
).build().convert(cfg)

with torch.device("meta"):
    model = cfg.build()
model.to_empty(device="cpu")

non_lora_trainable = [
    (n, tuple(p.shape))
    for n, p in model.named_parameters()
    if p.requires_grad and ".lora_a." not in n and ".lora_b." not in n
]
lora_trainable = sum(
    1 for n, _ in model.named_parameters()
    if ".lora_a." in n or ".lora_b." in n
)

print(f"LoRA tensors (expected trainable):     {lora_trainable}")
print(f"Non-LoRA params still trainable (BUG): {len(non_lora_trainable)}")
for name, shape in non_lora_trainable:
    print(f"  {name}  {shape}")
```

#### Output

```
LoRA tensors (expected trainable):     48
Non-LoRA params still trainable (BUG): 14

  tok_embeddings.weight  (2048, 256)
  layers.0.attention_norm.weight  (256,)
  layers.0.ffn_norm.weight  (256,)
  layers.1.attention_norm.weight  (256,)
  layers.1.ffn_norm.weight  (256,)
  layers.2.attention_norm.weight  (256,)
  layers.2.ffn_norm.weight  (256,)
  layers.3.attention_norm.weight  (256,)
  layers.3.ffn_norm.weight  (256,)
  layers.4.attention_norm.weight  (256,)
  layers.4.ffn_norm.weight  (256,)
  layers.5.attention_norm.weight  (256,)
  layers.5.ffn_norm.weight  (256,)
  norm.weight  (256,)
```

Every layer's two RMSNorm `weight`s plus the top-level `norm.weight`
plus the (2048×256) `tok_embeddings.weight` are all marked
`requires_grad=True`. Production flavors leak proportionally more.

### Expected behaviour

After `LoRAConverter` runs and the model is built, **only**
`.lora_a.*` / `.lora_b.*` parameters should be trainable.

### Environment

- torch: `2.12.0+cu126`
- Python: 3.12

**Comment by tianyu-l** [comment]
Oh this is definitely a bug, thanks for catching.

Maybe the fix should be:
Here we should traverse all modules, not just `Linear`s.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/lora.py#L201

Are you interested in submitting a PR to fix?

**Comment by yh8899** [comment]
Yes, i can try.

**Comment by yh8899** [comment]
> Oh this is definitely a bug, thanks for catching.
> 
> Maybe the fix should be: Here we should traverse all modules, not just `Linear`s. https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/lora.py#L201
> 
> Are you interested in submitting a PR to fix?

We can't just swap traverse(Linear.Config) for traverse(Module.Config). traverse matches on isinstance(val, config_cls) and, on a match, yields the node and stops descending into it (it's an if/elif: only un-matched nodes get recursed into). Can we freeze model in Trainer? That not need change the Lora.

**Comment by tianyu-l** [comment]
I see.

> Can we freeze model in Trainer? That not need change the Lora.

Ideally we only work in the config space, not actual model space.

> traverse matches on isinstance(val, config_cls) and, on a match, yields the node and stops descending into it (it's an if/elif: only un-matched nodes get recursed into).

Do you think we can add an kwarg `recurse` to `traverse` and use `True` for LoRA?

**Comment by yh8899** [comment]
> Do you think we can add an kwarg `recurse` to `traverse` and use `True` for LoRA?

This modification has a relatively wide range of impact. I will try it.
